### PR TITLE
Deploy devices using cli tool

### PR DESCRIPTION
**Pull request by fbeltrao**
Add gateway provisioning to cli tool

# PR for issue #1042

## What is being addressed

- Add capability to cli tool to provision gateway devices
- Use cli tool in Bicep deployment for device provisioning

## How is this addressed

Following changes were made to cli-tool:
- Adding gateway provisioning capability
- Add scripts to create standalone versions for Linux, Windows and Mac
- Add option to use environment variables to pass configuration (iothub connection string)

Following changes were made in Bicep template:
- Use lora-cli in Bicep template to provision gateway, concentrator and leaf devices

### CODE REVIEW

File: Tools/Cli-LoRa-Device-Provisioning/CommandLineOptions/AddGatewayOption.cs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+            )]
+        public string TwinHostAddress { get; set; }
+
+        [Option(
+            "network",
+            Required = false,
+            Default = DefaultNetwork,
+            HelpText = "Network name"
+            )]
+        public string Network { get; set; }
```
Comment (danigian) [inline Tools/Cli-LoRa-Device-Provisioning/CommandLineOptions/AddGatewayOption.cs:77]: This network is related to the discovery, correct? Can we adapt the HelpText to make it clearer? Something like "Network identifier for LNS Discovery purposes"

File: Tools/Cli-LoRa-Device-Provisioning/BuildForRelease.ps1 [hunk 2]
```diff
@@ -0,0 +1,33 @@
+﻿# Builds the cli and prepares it to be updated to a release
```
Comment (danigian) [inline Tools/Cli-LoRa-Device-Provisioning/BuildForRelease.ps1:1]: I love the use of the emojis in this .ps1 👍

File: TemplateBicep/main.bicep [hunk 3]
```diff
@@ -39,6 +39,9 @@ param gitUsername string = 'Azure'
 @description('The Git version to use. Default is 2.2.0.')
 param version string = '2.2.0'
 
+@description('The location of the cli tool to be used for device provisioning.')
+param loraCliUrl string = 'https://github.com/Azure/iotedge-lorawan-starterkit/releases/download/v${version}/lora-cli.linux-musl-x64.tar.gz'
```
Comment (danigian) [inline TemplateBicep/main.bicep:43]: Why defaulting to musl? Any specific reason?

**Review by danigian — COMMENTED** [review]
A couple comments so far. Can't currently test as I have not a ready setup on my Azure subscription.
As I don't want to block this, it would be great if others could also review this

File: TemplateBicep/README.md [hunk 4]
```diff
@@ -39,12 +39,13 @@ For device creation debugging, there are alternatives other than deploying the w
 Option 1: Run bash script locally
 
 ```plain
-FACADE_SERVER_URL="myapp.com/api" IOTHUB_NAME="<iothub-name>" RESOURCE_GROUP="<resource-group-name>" EDGE_GATEWAY_NAME="<iotedge-device-name>" STATION_DEVICE_NAME="<concentrator-device-name>" DEPLOY_DEVICE=1 ./create_device.sh
+FACADE_SERVER_URL="https://myapp.com/api" IOTHUB_CONNECTION_STRING="<iothub-connection-string>" LORA_CLI_URL="https://github.com/Azure/iotedge-lorawan-starterkit/releases/download/v2.2.0/lora-cli.linux-x64.tar.gz" EDGE_GATEWAY_NAME="<iotedge-device-name>" STATION_DEVICE_NAME="<concentrator-device-name>" DEPLOY_DEVICE=1 RESET_PIN=25 ./create_device.sh
```
Comment (Mandur) [inline TemplateBicep/README.md:42]: I thought the facade server url needed a trailing / after API. I will test this setting

Comment (Mandur) [reply TemplateBicep/README.md:42]: Sorr I thought this was the network server ;)

File: TemplateBicep/create_device.sh [hunk 5]
```diff
@@ -1,140 +1,47 @@
...
-        MODULE_CONFIG=$(cat edgeGatewayManifest.json)
-    fi
-
-    # Replace place holders in module configuration with values
-    local spiDevValue=''
-    if [ ! -z "${SPI_DEV}" ] && [ "$SPI_DEV" != "0" ]; then
-        spiDevValue=",\"SPI_DEV\": { \"value\": \"$SPI_DEV\" }"
-        echo "Setting SPI_DEV to $SPI_DEV"
+resolve_lora_region() {
+    local inputRegion=$1
```
Comment (Mandur) [inline TemplateBicep/create_device.sh:2]: As follow up work, could we actually don't do this mapping here and ask the users to give "EU863" so we could actually provision any region supported by the provisioning tool?

File: Tools/Cli-LoRa-Device-Provisioning/CommandLineOptions/AddGatewayOption.cs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+{
+    using System;
+    using CommandLine;
+
+    [Verb("add-gateway", HelpText = "Add a new gateway device to IoT Hub.")]
+    public class AddGatewayOption
+    {
+        internal const int DefaultSpiSpeed = 8;
+        internal const int DefaultSpiDev = 0;
+        internal const string DefaultAzureIotEdgeVersion = "1.4";
```
Comment (Mandur) [inline Tools/Cli-LoRa-Device-Provisioning/CommandLineOptions/AddGatewayOption.cs:14]: We would need to add this occurence to the playbook to not forget to update it when we updat iotedge version

File: Tools/Cli-LoRa-Device-Provisioning/CommandLineOptions/AddGatewayOption.cs [hunk 7]
```diff
@@ -0,0 +1,117 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace LoRaWan.Tools.CLI.Options
+{
+    using System;
+    using CommandLine;
+
+    [Verb("add-gateway", HelpText = "Add a new gateway device to IoT Hub.")]
+    public class AddGatewayOption
```
Comment (Mandur) [inline Tools/Cli-LoRa-Device-Provisioning/CommandLineOptions/AddGatewayOption.cs:10]: We have been avoiding the term gateway lately, what do you think about naming this "Add iot-edge"?

File: Tools/Cli-LoRa-Device-Provisioning/CommandLineOptions/AddGatewayOption.cs (hunk 1 above)
Comment (fbeltrao) [reply Tools/Cli-LoRa-Device-Provisioning/CommandLineOptions/AddGatewayOption.cs:77]: This parameter is optional, used to tag the device in IoT Hub.
Like this:
image

File: TemplateBicep/main.bicep (hunk 3 above)
Comment (fbeltrao) [reply TemplateBicep/main.bicep:43]: We use the tool within the [az-cli docker image](https://github.com/Azure/azure-cli/blob/dev/Dockerfile). It is built on top of Alpine. Running "regular" Linux was throwing errors due to missing libraries.

File: TemplateBicep/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (fbeltrao) [reply TemplateBicep/README.md]: Makes sense. Added.

File: Tools/Cli-LoRa-Device-Provisioning/CommandLineOptions/AddGatewayOption.cs (hunk 7 above)
Comment (fbeltrao) [reply Tools/Cli-LoRa-Device-Provisioning/CommandLineOptions/AddGatewayOption.cs:10]: Sure, I wasn't aware.
What is the name we use to describe to users? Iot edge device? Network server device?

File: Tools/Cli-LoRa-Device-Provisioning/CommandLineOptions/AddGatewayOption.cs (hunk 1 above)
Comment (fbeltrao) [reply Tools/Cli-LoRa-Device-Provisioning/CommandLineOptions/AddGatewayOption.cs:77]: But now, understand. You are right. Changed made.

File: TemplateBicep/create_device.sh (hunk 5 above)
Comment (fbeltrao) [reply TemplateBicep/create_device.sh:2]: What should the options be?
- EU863
- US902
- AS923-1
- AS923-2
- AS923-3
- CN470RP1
- CN470RP2

Comment (Mandur) [reply TemplateBicep/create_device.sh:2]: Yes and after #1925 gets merged AU915 as well

File: TemplateBicep/main.bicep (hunk 3 above)
Comment (danigian) [reply TemplateBicep/main.bicep:43]: Makes sense, thanks for the context

File: TemplateBicep/create_device.sh (hunk 5 above)
Comment (fbeltrao) [reply TemplateBicep/create_device.sh:2]: Will tackle in next PR

File: Tools/Cli-LoRa-Device-Provisioning/CommandLineOptions/AddGatewayOption.cs (hunk 7 above)
Comment (fbeltrao) [reply Tools/Cli-LoRa-Device-Provisioning/CommandLineOptions/AddGatewayOption.cs:10]: As discussed tags will be:

- lora_device_type: an array containing
  - network_server
  - basics_station
  - leaf
  - concentrator

Verb will be:
add-gateway

File: Tools/Cli-LoRa-Device-Provisioning/CommandLineOptions/AddGatewayOption.cs (hunk 6 above)
Comment (fbeltrao) [reply Tools/Cli-LoRa-Device-Provisioning/CommandLineOptions/AddGatewayOption.cs:14]: Will be automated here #1928

**Review by Mandur — APPROVED** [review]
Very cool, thank you very much for this!

**Pull request merged**